### PR TITLE
#169258639 notifications of new booking and travel request

### PR DESCRIPTION
**Pull request by hezronkimutai**
#### What does this PR do?
- This PR adds a functionality that will enable the travel admin to receive notification when a booking is made or the manager to receive the notification when a new trip request is made.
- This PR also adds a functionality in which users can mark notifications as read
#### Description of Task to be completed?
- Created two two routes, one for marking the all notifications as read and the other one for marking a single notification as read
- use socket IO to send notifications to the clients
- Add a user ID column in the accommodation table and bookingId in the notification table
#### How should this be manually tested?
- Open your terminal
- `git clone https://github.com/andela/team-odd-bn-backend.git`
- `cd team-odd-bn-backend`
- `git checkout ft-notification-booking-travel-request-169258639`
- Make sure your `.env` file is updated.
- In the browser, open `localhost:3000/static` (make sure you frequently reload it as the tests are running)
- Run `npm test` to run the tests and view the browser's console.
- Open postman and test the endpoints`api/v1/notifications/markRead` and `api/v1/notifications/markRead/:notificationId`
- You can also test if notifications are sent to the client from postman by creating a new trip or making a booking
#### Any background context you want to provide?
**The input data**
`
{
   "notificationIds":
   [1,2,3,4,5,11]
}
`
#### What are the relevant pivotal tracker stories?
[#169258639](https://www.pivotaltracker.com/n/projects/2407428/stories/169258639)
#### Screenshots (if appropriate)
**Data sent to the client**
Screen Shot 2019-12-11 at 17 08 57

**Successful mark notifications as read**
Screen Shot 2019-12-11 at 17 17 22

**String notification ID provided in the array**
Screen Shot 2019-12-11 at 17 18 16

**String notification ID provided in the params**
Screen Shot 2019-12-11 at 17 22 39

#### Questions:

### CODE REVIEW

File: src/controllers/BookingController.js [hunk 1]
```diff
@@ -18,8 +19,12 @@ class BookingController {
   static async bookAccommodation(req, res) {
     try {
       const result = await AccommodationService.createAccommodationService(req);
+      req.result = result;
+      await NotificationService.newBookingNotification(req);
       return Response.successMessage(req, res, 'You have booked an accommodation facility successfully', result, 201);
     } catch (error) {
+      console.log(error);
+
```
Comment (victkarangwa) [inline src/controllers/BookingController.js]: Can we paint this catch error in a smart way?
like:
```js
return Response.errorMessage(req, res, error.message, 500);
```

File: src/helpers/TripHelper.js [hunk 2]
```diff
@@ -34,9 +34,14 @@ class TripHelper {
...
         });
       });
+
+      req.result = newTrip;
+      await NotificationService.newTripRequestNotification(req);
       emailHelper.approveEmailHelper(req, process.env.MANAGER_EMAIL);
       return Response.successMessage(req, res, 'Trip requested successfully', newTrip, 201);
     } catch (err) {
+      console.log(err);
+
```
Comment (victkarangwa) [inline src/helpers/TripHelper.js]: The same here

File: src/services/NotificationService.js [hunk 3]
```diff
@@ -38,18 +42,116 @@ class NotificationService {
...
+   * @returns {object} data
+   */
+  static async saveNotification(dataNotification) {
+    const newNotification = await CommonQueries.create(notifications, dataNotification);
+    return newNotification;
+  }
+
+  /**
+   * generate notification
+   * @static
```
Comment (victkarangwa) [inline src/services/NotificationService.js]: I think it should reflect the functionality like :
```js
  /**
   * Mark notification as read

**Review by victkarangwa — COMMENTED** [review]
You did a fantastic job @hezronkimutai

**Review by victkarangwa — APPROVED** [review]
@hezronkimutai congrats for implementing the feedback

File: src/public/index.html [hunk 4]
```diff
@@ -14,8 +14,12 @@
 <script src="/socket.io/socket.io.js"></script>
 <script>
   const socket = io('http://localhost:3000');
-  socket.on('approve_reject', (name) => {
+  const rr = ['approve_reject_client','booking_client','trip_request_client'];
+  rr.forEach((ev)=>{
+    socket.on(ev, (name) => {
```
Comment (william000000) [inline src/public/index.html]: It is good but It should be better when you rename this variable `rr` to a meaningful name.

File: src/services/LocationService.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+   */
+class LocationService {
+  /**
+     * user edit trips
+     * @static
+     * @param {object} req pass request body
+     * @memberof TripService
+     * @returns {object} either an error or data
+     */
+  static async getLocations() {
```
Comment (william000000) [inline src/services/LocationService.js]: Rename ` @memberof TripService` to ` @memberof LocationService`

File: src/controllers/NotificationController.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+   * @param {object} res response object
+   * @memberof AccommodationController
+   * @returns {object} data
+   */
+  static async markNotificationsAsRead(req, res) {
+    try {
+      const { notificationIds } = req.body;
+      const notificationArr = req.params.notificationId
+        ? notificationIds.filter(id => id === parseInt(req.params.notificationId, 10))
+        : notificationIds;
```
Comment (NiyoEric) [inline src/controllers/NotificationController.js]: I recommend you use ES6 to destruct object from `req` so that code looks good.

**Review by william000000 — COMMENTED** [review]
You have done a great job @hezronkimutai, Keep it up. However, you can check out the comments I left to beautify your work.

File: src/routes/api/notificationRoute.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ *           notificationIds: [1,2,3,4,5]
+ *
+ */
+notificationRoute
+  .patch(
+    '/markRead',
+    Validate.validateNotificationIdRules(),
+    checkInputDataError,
+    verifyToken,
+    isUserVerified,
```
Comment (NiyoEric) [inline src/routes/api/notificationRoute.js]: I suggest you change the order of middlewares. For example `isUserVerified` comes first.

File: src/routes/api/notificationRoute.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+ *           notificationIds: [1,2,3,4,5]
+ *
+ */
+notificationRoute
+  .patch(
+    '/markRead/:notificationId',
+    Validate.validateNotificationIdRules(),
+    checkInputDataError,
+    verifyToken,
+    isUserVerified,
```
Comment (NiyoEric) [inline src/routes/api/notificationRoute.js]: I suggest you `isUserVerified` comes first.

**Review by NiyoEric — CHANGES_REQUESTED** [review]
Good job @hezronkimutai! I liked how you delivered the task. Also, please incorporates the feedback I left.

**Review by NiyoEric — APPROVED** [review]

**Review by william000000 — APPROVED** [review]

**Review by julietezekwe — APPROVED** [review]

**Pull request merged**